test/runnable/testassert.d

```diff
@@ -112,19 +119,54 @@ void test20375() @safe
 
     {
         auto a = RefCounted.create();
-        assert(a == RefCounted.create());
+        RefCounted.instances++; // we're about to construct an instance below, increment manually
+        assert(a == RefCounted()); // both operands are pure expressions => no temporaries
     }
 
     assert(RefCounted.instances == 0);
+    assert(RefCounted.postblits == 0);
 
     {
         auto a = RefCounted.create();
-        const msg = getMessage(assert(a != RefCounted.create()));
-        // assert(msg == "RefCounted() == RefCounted()"); // Currently not formatted
-        assert(msg == "assert(a != RefCounted.create()) failed");
+        assert(a == RefCounted.create()); // impure rhs is promoted to a temporary lvalue => copy for a.opEquals(rhs)
     }
 
     assert(RefCounted.instances == 0);
+    assert(RefCounted.postblits == 1);
+    RefCounted.postblits = 0;
+
+    {
+        const msg = getMessage(assert(RefCounted.create() != RefCounted.create())); // both operands promoted to temporary lvalues
+        assert(msg == "RefCounted() == RefCounted()");
+    }
+
+    version (DMD_Win64) // FIXME: temporaries apparently not destructed when unwinding via AssertError
```

Comment:
I'm mildly worried about a FIXME targeting stable, is tis problem EH, ABI, Codegen, something else?

Reply:
I think much of these tests here are actually testing undefined behavior, as the lifetime of temporaries in statements throwing an AssertError is tracked, continuing in the tests after catching the `AssertError`, checking its message and checking that the temporaries have been properly destructed.

I'm unsure about the spec. Win64 with DMD seems to be the only platform eliding the cleanups, so I found it mildly worrying too, hence the comment, but at the same time I don't care about the potential DMD-specific codegen or druntime EH issue, as it's definitely unrelated to this PR.

Reply:
Fair enough.